roachtest: tpchvec/perf_no_stats failed

[(roachtest).tpchvec/perf_no_stats failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2114210&tab=buildLog) on [master@bfa6307c292ef4dfed4a53cb99f506e6dab26533](https://github.com/cockroachdb/cockroach/commits/bfa6307c292ef4dfed4a53cb99f506e6dab26533):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/tpchvec/perf_no_stats/run_1
	test_runner.go:804: test timed out (10h0m0s)

```

<details><summary>More</summary><p>


Artifacts: [/tpchvec/perf_no_stats](https://teamcity.cockroachdb.com/viewLog.html?buildId=2114210&tab=artifacts#/tpchvec/perf_no_stats)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Atpchvec%2Fperf_no_stats.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

Query 18:
```
panic: interface conversion: tree.Datum is *tree.DDecimal, not *tree.DFloat

goroutine 29192 [running]:
panic(0x45f95a0, 0xc0089530b0)
	/usr/local/go/src/runtime/panic.go:1064 +0x46d fp=0xc000a2fb30 sp=0xc000a2fa78 pc=0x7b3a5d
runtime.panicdottypeE(...)
	/usr/local/go/src/runtime/iface.go:260
runtime.panicdottypeI(0x5757120, 0x4a39a20, 0x48f7dc0)
	/usr/local/go/src/runtime/iface.go:270 +0xcb fp=0xc000a2fb58 sp=0xc000a2fb30 pc=0x78928b
github.com/cockroachdb/cockroach/pkg/sql/sem/builtins.(*floatSumAggregate).Add(0xc00a8b8240, 0x56c6020, 0xc005b5e100, 0x5757120, 0xc00ad0ad80, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/builtins/aggregate_builtins.go:1869 +0xc7 fp=0xc000a2fb88 sp=0xc000a2fb58 pc=0x18b9317
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*aggregatorBase).accumulateRowIntoBucket(0xc001be2800, 0xc005bc08c8, 0x6, 0x6, 0x0, 0x0, 0x0, 0xc007732d20, 0x6, 0x6, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/aggregator.go:834 +0x3e5 fp=0xc000a2fcf0 sp=0xc000a2fb88 pc=0x1ce05b5
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*orderedAggregator).accumulateRow(0xc001be2800, 0xc005bc08c8, 0x6, 0x6, 0xc005bc08c8, 0x6)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/aggregator.go:895 +0xbe fp=0xc000a2fd60 sp=0xc000a2fcf0 pc=0x1ce0eae
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*orderedAggregator).accumulateRows(0xc001be2800, 0xc005b5e100, 0x574dcc0, 0xc00086d5c0, 0xc001be2800, 0x56c6020)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/aggregator.go:559 +0x222 fp=0xc000a2fdf8 sp=0xc000a2fd60 pc=0x1cdedd2
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*orderedAggregator).Next(0xc001be2800, 0x56c6020, 0xc005b5e100, 0x4b63bb8, 0x8)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/aggregator.go:752 +0x14c fp=0xc000a2fe60 sp=0xc000a2fdf8 pc=0x1ce00dc
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*sortTopKProcessor).Start(0xc005be3400, 0x56c6020, 0xc005b5e100, 0x56c60e0, 0xc004585380)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/sorter.go:395 +0xed fp=0xc000a2ff40 sp=0xc000a2fe60 pc=0x1d182ed
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).Run(0xc005be3400, 0x56c6020, 0xc005b5e100)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:747 +0x52 fp=0xc000a2ff80 sp=0xc000a2ff40 pc=0x1af7422
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).startInternal.func1(0xc001cb1680, 0x5, 0xa, 0x56c6020, 0xc005b5e100, 0xc004f7c360, 0x4)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:316 +0x5c fp=0xc000a2ffa8 sp=0xc000a2ff80 pc=0x1cd8dbc
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1373 +0x1 fp=0xc000a2ffb0 sp=0xc000a2ffa8 pc=0x7e70c1
created by github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).startInternal
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:315 +0x2d9
```

The timeout also occurred on query 18, but all the nodes remained healthy, and I don't know why it would stall.

--- a/pkg/sql/distsql_physical_planner.go
+++ b/pkg/sql/distsql_physical_planner.go
@@ -2385,8 +2385,9 @@ func (dsp *DistSQLPlanner) createPlanForJoin(
 
 	leftMap, rightMap := leftPlan.PlanToStreamColMap, rightPlan.PlanToStreamColMap
 	helper := &joinPlanningHelper{
-		numLeftCols:             n.pred.numLeftCols,
-		numRightCols:            n.pred.numRightCols,
+		numLeftOutCols:          n.pred.numLeftCols,
+		numRightOutCols:         n.pred.numRightCols,
+		numAllLeftCols:          len(leftPlan.ResultTypes),
 		leftPlanToStreamColMap:  leftMap,
 		rightPlanToStreamColMap: rightMap,
 	}

--- a/pkg/sql/distsql_plan_join.go
+++ b/pkg/sql/distsql_plan_join.go
@@ -154,8 +154,9 @@ func (dsp *DistSQLPlanner) createPlanForInterleavedJoin(
 
 	leftMap, rightMap := plans[0].PlanToStreamColMap, plans[1].PlanToStreamColMap
 	helper := &joinPlanningHelper{
-		numLeftCols:             len(n.left.cols),
-		numRightCols:            len(n.right.cols),
+		numLeftOutCols:          len(n.left.cols),
+		numRightOutCols:         len(n.right.cols),
+		numAllLeftCols:          len(plans[0].ResultTypes),
 		leftPlanToStreamColMap:  leftMap,
 		rightPlanToStreamColMap: rightMap,
 	}
@@ -283,7 +284,14 @@ func (dsp *DistSQLPlanner) createPlanForInterleavedJoin(
 // joinPlanningHelper is a utility struct that helps with the physical planning
 // of joins.
 type joinPlanningHelper struct {
-	numLeftCols, numRightCols                       int
+	// numLeftOutCols and numRightOutCols store the number of columns that need
+	// to be included in the output of the join from each of the sides.
+	numLeftOutCols, numRightOutCols int
+	// numAllLeftCols stores the width of the rows coming from the left side.
+	// Note that it includes all of the left "out" columns and might include
+	// other "internal" columns that are needed to merge the streams for the
+	// left input.
+	numAllLeftCols                                  int
 	leftPlanToStreamColMap, rightPlanToStreamColMap []int
 }
 
@@ -302,16 +310,16 @@ func (h *joinPlanningHelper) joinOutColumns(
 	}
 
 	// The join columns are in two groups:
-	//  - the columns on the left side (numLeftCols)
-	//  - the columns on the right side (numRightCols)
-	for i := 0; i < h.numLeftCols; i++ {
+	//  - the columns on the left side (numLeftOutCols)
+	//  - the columns on the right side (numRightOutCols)
+	for i := 0; i < h.numLeftOutCols; i++ {
 		joinToStreamColMap[i] = addOutCol(uint32(h.leftPlanToStreamColMap[i]))
 	}
 
-	if joinType != sqlbase.LeftSemiJoin && joinType != sqlbase.LeftAntiJoin {
-		for i := 0; i < h.numRightCols; i++ {
-			joinToStreamColMap[h.numLeftCols+i] = addOutCol(
-				uint32(h.numLeftCols + h.rightPlanToStreamColMap[i]),
+	if joinType.ShouldIncludeRightColsInOutput() {
+		for i := 0; i < h.numRightOutCols; i++ {
+			joinToStreamColMap[h.numLeftOutCols+i] = addOutCol(
+				uint32(h.numAllLeftCols + h.rightPlanToStreamColMap[i]),
 			)
 		}
 	}
@@ -319,7 +327,7 @@ func (h *joinPlanningHelper) joinOutColumns(
 	return post, joinToStreamColMap
 }
 
-// remapOnExpr remaps ordinal references in the on condition (which refer to the
+// remapOnExpr remaps ordinal references in the ON condition (which refer to the
 // join columns as described above) to values that make sense in the joiner (0
 // to N-1 for the left input columns, N to N+M-1 for the right input columns).
 func (h *joinPlanningHelper) remapOnExpr(
@@ -329,17 +337,17 @@ func (h *joinPlanningHelper) remapOnExpr(
 		return execinfrapb.Expression{}, nil
 	}
 
-	joinColMap := make([]int, h.numLeftCols+h.numRightCols)
+	joinColMap := make([]int, h.numLeftOutCols+h.numRightOutCols)
 	idx := 0
 	leftCols := 0
-	for i := 0; i < h.numLeftCols; i++ {
+	for i := 0; i < h.numLeftOutCols; i++ {
 		joinColMap[idx] = h.leftPlanToStreamColMap[i]
 		if h.leftPlanToStreamColMap[i] != -1 {
 			leftCols++
 		}
 		idx++
 	}
-	for i := 0; i < h.numRightCols; i++ {
+	for i := 0; i < h.numRightOutCols; i++ {
 		joinColMap[idx] = leftCols + h.rightPlanToStreamColMap[i]
 		idx++
 	}

--- a/pkg/sql/distsql_spec_exec_factory.go
+++ b/pkg/sql/distsql_spec_exec_factory.go
@@ -954,8 +954,9 @@ func (e *distSQLSpecExecFactory) constructHashOrMergeJoin(
 	resultColumns := getJoinResultColumns(joinType, leftPhysPlan.ResultColumns, rightPhysPlan.ResultColumns)
 	leftMap, rightMap := leftPhysPlan.PlanToStreamColMap, rightPhysPlan.PlanToStreamColMap
 	helper := &joinPlanningHelper{
-		numLeftCols:             len(leftPhysPlan.ResultColumns),
-		numRightCols:            len(rightPhysPlan.ResultColumns),
+		numLeftOutCols:          len(leftPhysPlan.ResultTypes),
+		numRightOutCols:         len(rightPhysPlan.ResultTypes),
+		numAllLeftCols:          len(leftPhysPlan.ResultTypes),
 		leftPlanToStreamColMap:  leftMap,
 		rightPlanToStreamColMap: rightMap,
 	}

--- a/pkg/sql/opt/exec/execbuilder/testdata/distsql_join
+++ b/pkg/sql/opt/exec/execbuilder/testdata/distsql_join
@@ -197,3 +197,47 @@ https://cockroachdb.github.io/distsqlplan/decode.html#eJy8ll1v-jYUh-_3KaxzBatRsB
 # SELECT url FROM [EXPLAIN (DISTSQL) (SELECT l.k, r.k FROM (SELECT * FROM distsql_mj_test ORDER BY k) l INNER JOIN (SELECT * FROM distsql_mj_test ORDER BY k) r ON l.k = r.k)]
 # ----
 # https://cockroachdb.github.io/distsqlplan/decode.html#eJyskk9Lw0AQxe9-CpmT0pVk06SHgJBrBVup3qSUmB3jSpqNsxtQSr-7JHtIE8z2D952Z-f35r1ldlAqgYt0ixriV-DAIII1g4pUhlorasq2aS6-IfYZyLKqTVNeM8gUIcQ7MNIUCDEs1J2qvAgYCDSpLNq2PQNVmw7SJs0R4tmeHQhzt_BL-lbgClOB5Pk9eahIblP6SYTURn8Vm-3nxqA2wGBZm_g64TDmgJ_j4FmRQfJ4f3jCJ6PywaUBg_8KOL0g4PT0gOGofKdal4oEEorhThxv-cPjI1KOD0qWSF7YN1rgu7lJ-OT2nmT-YY_dF7EkGI0R9WIc2fAV6kqVGk9acb_JgCJH-yda1ZThE6msHWOvy5ZrCwK1sa8ze5mX9qkxeAhzJxy44cAJRz2YD-GpEw7dk8MzJgdDOHLC_mDyen_1GwAA___yFZqv
+
+# Regression test for incorrectly populating PlanToStreamColMap for the right
+# side of merge joins (#51883).
+statement ok
+CREATE TABLE customer (c_custkey INT8 PRIMARY KEY);
+CREATE TABLE orders (
+  o_orderkey INT8 PRIMARY KEY, o_custkey INT8 NOT NULL,
+  CONSTRAINT orders_fkey_customer FOREIGN KEY (o_custkey) REFERENCES customer (c_custkey)
+);
+CREATE TABLE lineitem (
+  l_orderkey INT8 NOT NULL,
+  l_linenumber INT8 NOT NULL,
+  l_quantity FLOAT8 NOT NULL,
+  PRIMARY KEY (l_orderkey, l_linenumber),
+  CONSTRAINT lineitem_fkey_orders FOREIGN KEY (l_orderkey) REFERENCES orders (o_orderkey)
+);
+ALTER TABLE lineitem SPLIT AT SELECT i FROM generate_series(1, 9) AS g(i);
+ALTER TABLE lineitem EXPERIMENTAL_RELOCATE SELECT ARRAY[i%5+1], i FROM generate_series(0, 9) AS g(i)
+
+# This query checks that there is no type mismatch during the vectorized flow
+# setup which would occur if the stream column indices got messed up.
+statement ok
+SET vectorize=on;
+EXPLAIN (VEC)
+  SELECT
+    sum(l_quantity)
+  FROM
+    customer,
+    orders
+    INNER MERGE JOIN lineitem ON o_orderkey = l_orderkey
+  WHERE
+    o_orderkey
+    IN (
+        SELECT
+          l_orderkey
+        FROM
+          lineitem
+        GROUP BY
+          l_orderkey
+        HAVING
+          sum(l_quantity) > 300
+      )
+    AND c_custkey = o_custkey;
+RESET vectorize

